### PR TITLE
TMNF BC adapter; remove do_pretrain (issue #395)

### LINKED ISSUE #395: BC refactor — Phase 3: TMNF BCAdapter (SimplePolicy source) and drop `do_pretrain`
Refs #392 (parent). Depends on #393 (phase 1) and #394 (phase 2).

Wire TMNF onto the framework BC seam using its existing in-game `SimplePolicy` demonstration source, and **remove the legacy `do_pretrain: true` path** as a breaking change.

This is the "dedicated TMNF code" piece of the refactor. A separate subissue (#396 — TMNF .Gbx replay ingest) adds the replay-file source.

## Scope — TMNF BCAdapter

- New `games/tmnf/bc.py` (or `games/tmnf/bc_adapter.py`) implementing `BCAdapter`:
  - `supported_targets = ("hill_climbing",)` initially. `default_target = "hill_climbing"`.
  - `validate_replay_dir(replay_dir, *, race)` — synthetic info object when `replay_dir is None`, since this source is live demonstrations rather than replay files. Document that "race" is unused for TMNF.
  - `build_dataset(replay_dir, save_path, *, obs_spec, player_id, race, max_replays, training_params)` — drives `SimplePolicy` in the TMNF env for `N_DEMO_LAPS` (or `max_replays` reinterpreted as "n_demo_laps") and writes a `demos.npz` compatible with the framework loader. This is the loop currently in `rl/pretrain.py:collect_demos` (lines 18-33), refactored to emit the standard NPZ format.
  - `fit_bc(dataset, obs_spec, *, target, training_params)` — for `target="hill_climbing"`, runs the three-head `np.linalg.lstsq` fit currently in `rl/pretrain.py:fit_weighted_linear` (lines 36-51) and returns a `WeightedLinearPolicy`.
- Wire `TMNFAdapter.bc = TMNFBCAdapter()`.

## Scope — drop `do_pretrain` (breaking change)

- Delete the `do_pretrain` branch in `framework/training.py` (lines 1790-1859 region — the `_will_pretrain` block and the `from rl.pretrain import run as _pretrain_run` import).
- Remove `do_pretrain` from `framework/run_config.py` (lines 50, 66).
- Delete `rl/pretrain.py` outright (no shim — `grep -r "from rl.pretrain" .` confirms only `framework/training.py` consumes it, and that consumer is removed).
- Update `games/tmnf/config/training_params.yaml` to drop the `do_pretrain` key if present.
- Bump **minor** version (per CLAUDE.md SemVer policy — config-shape breaking change) via `scripts/release.py` after merge; flag in CHANGELOG as `### Breaking`.

## Migration note (CHANGELOG + README)

> `do_pretrain: true` in `training_params.yaml` is no longer honoured. Use `python main.py <experiment> --game tmnf --bc` (no `--replay-dir` for the SimplePolicy source) to reproduce the same WeightedLinearPolicy lstsq warm-start.

## Tests

- `tests/test_tmnf_bc.py` (new): TMNF BCAdapter end-to-end against a stubbed `TMNFEnv` + `SimplePolicy` — assert `demos.npz` shape, `policy_weights.yaml` round-trips, `WeightedLinearPolicy` from BC produces the same fit as the legacy `rl/pretrain.fit_weighted_linear` given the same demos.
- Drop / replace any tests of `do_pretrain` / `rl/pretrain.run` (`grep -rn pretrain tests/`).
- Update `tests/README.md`.

## Acceptance

- `python main.py myrun --game tmnf --bc` produces a `policy_weights.yaml` indistinguishable (within tolerance) from what `do_pretrain: true` produced on the same seed.
- `framework/training.py` no longer imports from `rl/`.
- `CHANGELOG.md` `### Breaking` entry covering both the framework BC protocol arrival and the `do_pretrain` removal.
- `pre-commit run --all-files` clean.

### PR DESCRIPTION
**Pull request by espenhk**
Closes #395

## Summary

- Implements `games/tmnf/bc_adapter.py` — TMNF's BCAdapter using the in-game `SimplePolicy` as the demonstration source. Wired on `TMNFAdapter.bc` and added to `_build_bc_obs_spec` in `main.py` so `python main.py <experiment> --game tmnf --bc` is the supported way to reproduce the legacy `do_pretrain` warm-start.
- **Breaking:** removes the `do_pretrain: true` training-params knob and its `rl/pretrain.py` landing pad. `RunConfig.from_training_params` silently drops stale `do_pretrain` keys so legacy configs still load.
- 16 new tests in `tests/test_tmnf_bc.py` cover the protocol surface, lstsq round-trip, end-to-end orchestration with a stubbed env, and the `do_pretrain` removal guards.
- `.Replay.Gbx` replay-file ingest is **out of scope** — `validate_replay_dir` and `build_dataset` both raise a clear error pointing to follow-up issue #396.

## Related issues

Refs #392 (parent). Phases 1 + 2 landed via #403; phase 3b (`.Replay.Gbx` ingest) is #396; phase 4 (polish, version bump) is #397.

This PR **targets `feature/bc-refactor-392`**, not `main`. The feature branch will merge to `main` once #396 + #397 also land on it.

## Validation

```bash
PYTHONPATH=. poetry run python -m pytest tests/test_framework_bc.py tests/test_sc2_bc_adapter.py tests/test_sc2_replay_bc.py tests/test_tmnf_bc.py -v
PYTHONPATH=. poetry run python -m pytest tests/ \
    --ignore=tests/test_env_termination.py \
    --ignore=tests/test_grid_search.py \
    --ignore=tests/integration/
```

161 BC tests green (13 framework + 4 sc2_bc_adapter + 128 sc2_replay_bc + 16 tmnf_bc). Pre-existing `tests/test_sc2_analytics.py` failures (matplotlib stub mismatch) are unrelated and also reproduce on `feature/bc-refactor-392`.

## Feature details

- **User problem:** TMNF previously had a separate `do_pretrain: true` knob that pre-trained a `WeightedLinearPolicy` from `SimplePolicy` demonstrations via `rl/pretrain.py`. This was structurally different from SC2's framework-level BC pipeline, so the two games had two completely different BC stories.
- **Proposed solution:** Implement `BCAdapter` for TMNF (the same way `SC2BCAdapter` does for SC2), wire it on `TMNFAdapter.bc`, and add TMNF to the `_build_bc_obs_spec` dispatch in `main.py`. The user-facing UX collapses to `python main.py <experiment> --game tmnf --bc`. The same lstsq logic that lived in `rl/pretrain.py` is preserved verbatim inside `games/tmnf/bc_adapter.py`, just exposed through the framework BC pipeline.
- **Trade-offs / follow-ups:**
  - `.Replay.Gbx` ingest (the TMNF analogue of SC2's `.SC2Replay` parsing) is deferred to #396. The TMNF BCAdapter raises with a pointer when a `replay_dir` is provided, so users discover the limit early.
  - Only `bc_target: hill_climbing` is supported on TMNF today — the lstsq fit produces a `WeightedLinearPolicy`. Other targets (neural_dqn, reinforce, etc.) are not in scope for phase 3.
  - This PR carries the SemVer-minor-breaking `do_pretrain` removal. The bump happens in #397 (phase 4) once the feature branch is ready to merge to `main`.

## Refactor / docs / chore details

- **What changed:**
  - New `games/tmnf/bc_adapter.py` — TMNFBCAdapter (protocol impl), `collect_demos`, `fit_weighted_linear`, `_bc_residual_loss`, `_make_tmnf_env`, `_save_dataset_npz`.
  - `games/tmnf/adapter.py`: wired `TMNFBCAdapter()` instance to `self.bc`.
  - `main.py`: added TMNF dispatch to `_build_bc_obs_spec`.
  - `framework/training.py`: removed the `_will_pretrain` branch and the `from rl.pretrain import run` import.
  - `framework/run_config.py`: removed `do_pretrain` field and `from_training_params` entry.
  - `rl/pretrain.py`: deleted.
  - `games/tmnf/entry.py`, `grid_search.py`, `docs/framework/run_config.md`: removed `do_pretrain` references.
  - `CHANGELOG.md`: `Added` + `Breaking` entries under `## [Unreleased]`.
  - `CLAUDE.md`: `--bc` paragraph rewritten as game-agnostic.
  - `games/tmnf/README.md`: new "Behaviour cloning warm-start" section.
  - `tests/README.md`: new "Framework BC seam", "SC2 BC adapter", "TMNF BC" sections.
  - `tests/test_tmnf_bc.py`: 16 new tests.

- **Why this is safe:**
  - The lstsq fit in `fit_weighted_linear` is the same code that ran under `do_pretrain: true`; the on-disk `policy_weights.yaml` is byte-compatible (the weight format is `{name: float}` per head, fixed alongside the lift since the legacy code was passing a stale list format).
  - Stale `do_pretrain: true` in any user's `training_params.yaml` is silently dropped by `RunConfig.from_training_params` — no crash on legacy configs.
  - `.Replay.Gbx` directories raise a clear error from the adapter rather than silently doing nothing.
  - The fake-env test harness exercises the full pipeline without needing TMInterface + Trackmania.

https://claude.ai/code/session_017YYNh7Zfp7RQSAhLKYu9Qx

### CODE REVIEW

**Comment by espenhk** [comment]
Closes #393, closes #394
